flytekit/core/type_engine.py

```diff
@@ -258,27 +267,68 @@ def _serialize_flyte_type(self, python_val: T, python_type: Type[T]):
         """
         If any field inside the dataclass is flyte type, we should use flyte type transformer for that field.
         """
-        from flytekit.types.schema.types import FlyteSchema, FlyteSchemaTransformer
+        from flytekit.types.directory.types import FlyteDirectory
+        from flytekit.types.file import FlyteFile
+        from flytekit.types.schema.types import FlyteSchema
 
         for f in dataclasses.fields(python_type):
             v = python_val.__getattribute__(f.name)
-            if inspect.isclass(f.type) and issubclass(f.type, FlyteSchema):
-                FlyteSchemaTransformer().to_literal(FlyteContext.current_context(), v, f.type, None)
+            if inspect.isclass(f.type) and (
+                issubclass(f.type, FlyteSchema) or issubclass(f.type, FlyteFile) or issubclass(f.type, FlyteDirectory)
+            ):
+                TypeEngine.to_literal(FlyteContext.current_context(), v, f.type, None)
             elif dataclasses.is_dataclass(f.type):
                 self._serialize_flyte_type(v, f.type)
 
     def _deserialize_flyte_type(self, python_val: T, expected_python_type: Type["FlyteSchema"]):
+        from flytekit.types.directory.types import FlyteDirectory, FlyteDirToMultipartBlobTransformer
+        from flytekit.types.file.file import FlyteFile, FlyteFilePathTransformer
         from flytekit.types.schema.types import FlyteSchema, FlyteSchemaTransformer
 
         for f in dataclasses.fields(expected_python_type):
             v = python_val.__getattribute__(f.name)
-            if inspect.isclass(f.type) and issubclass(f.type, FlyteSchema):
-                t = FlyteSchemaTransformer()
-                t.to_python_value(
-                    FlyteContext.current_context(),
-                    Literal(scalar=Scalar(schema=Schema(v.remote_path, t._get_schema_type(f.type)))),
-                    f.type,
-                )
+            if inspect.isclass(f.type):
+                if issubclass(f.type, FlyteSchema):
+                    t = FlyteSchemaTransformer()
+                    t.to_python_value(
```

Comment:
I'm actually a bit confused now.  I didn't see this before.  We just throw away the output of the to_python_value call?  Why don't we use it?

My concern is that the value is going to be an object that's hooked up to some downloading function.  https://github.com/flyteorg/flytekit/blob/742f40eb718777887043db3842b076b7f87a8427/flytekit/types/schema/types.py#L230  Doesn't this function need to be invoked for the content to actually be there?  Same for file/directory.